lib/fluent/plugin/in_forward.rb

```diff
@@ -82,11 +83,15 @@ def shutdown
       @usock.close
       @thread.join
       @lsock.close
+      @connections.each { |c| c.close }
     end
 
     def listen
       log.info "listening fluent socket on #{@bind}:#{@port}"
-      s = Coolio::TCPServer.new(@bind, @port, Handler, @linger_timeout, log, method(:on_message))
+      s = Coolio::TCPServer.new(@bind, @port, Handler, @linger_timeout, log, method(:on_message)) do |connection|
+        @connections.reject! { |conn| conn.closed? }
```

Comment:
It looks needed for me too to guard `@connections.reject!` using Mutex.
I mean that:
```ruby
@mutex.synchronize do
  @connections.reject!{ |conn| conn.closed? }
  @connections << connection
end
```

Connection callback may run twice immediately if many connection arrived at a time.